Firestore/Example/Tests/SpecTests/FSTSpecTests.mm

```diff
@@ -330,10 +330,29 @@ - (SnapshotVersion)parseVersion:(NSNumber *_Nullable)version {
 }
 
 - (absl::optional<BloomFilter>)parseBloomFilter:(NSDictionary *_Nullable)bloomFilterProto {
-  // TODO(Mila): None of the ported spec tests actually has the bloom filter json string, so hard
-  // code a null value for now. Actual parsing code will be written in the next PR, where we can
-  // validate the parsing result.
-  return absl::nullopt;
+  if (bloomFilterProto == nil) {
+    return absl::nullopt;
+  }
+  NSDictionary *bitsData = bloomFilterProto[@"bits"];
+
+  // Decode base64 string into uint8_t vector. If bitmap is not specified in proto, use default
+  // empty string.
+  NSString *bitmapEncoded = bitsData[@"bitmap"];
+  std::string bitmapDecoded;
+  absl::Base64Unescape([bitmapEncoded cStringUsingEncoding:NSASCIIStringEncoding], &bitmapDecoded);
+  std::vector<uint8_t> bitmap(bitmapDecoded.begin(), bitmapDecoded.end());
+
+  // If not specified in proto, default padding and hashCount to 0.
+  int32_t padding = [bitsData[@"padding"] intValue];
+  int32_t hashCount = [bloomFilterProto[@"hashCount"] intValue];
+  StatusOr<BloomFilter> maybeBloomFilter = BloomFilter::Create(bitmap, padding, hashCount);
+
+  if (maybeBloomFilter.ok()) {
+    return std::move(maybeBloomFilter).ValueOrDie();
+  } else {
+    LOG_WARN("Parsing BloomFilterProto failed: %s", maybeBloomFilter.status().error_message());
```

Comment:
If parsing the bloom filter fails, this should fail the test. Can you modify this code so that the test that calls this method fails? Maybe call XCTFail() and see if that works?

Reply:
Some of our spec tests have invalid bloom filter proto, to test the flow "fail to create a BloomFilter -> trigger full re-query". So some of the bloom filter creation are doomed to fail, and the test should still carry on.

Reply:
Discussed offline. The bloom filter parsing logic will be moved to later in the stack, for parity with the web and android sdks and also so that the spec tests can exercise the logic that parses the bloom filter from the proto.